packages/nx/src/plugins/js/lock-file/pnpm-parser.ts

```diff
@@ -307,25 +318,28 @@ function findPackageName(
 
   // snapshot already has a name
   if (snapshot.name) {
-    return snapshot.name;
+    packageNames.add(snapshot.name);
   }
   // it'a a root dependency
   const rootDependencyName =
     matchedDependencyName(data.importers['.']) ||
     // only root importers have devDependencies
     matchPropValue(data.importers['.'].devDependencies);
   if (rootDependencyName) {
-    return rootDependencyName;
+    packageNames.add(rootDependencyName);
   }
   // find a snapshot that has a dependency that points to this snapshot
   const snapshots = Object.values(data.packages);
   for (let i = 0; i < snapshots.length; i++) {
     const dependencyName = matchedDependencyName(snapshots[i]);
     if (dependencyName) {
-      return dependencyName;
+      packageNames.add(dependencyName);
     }
   }
-  return extractNameFromKey(key);
+  if (packageNames.size === 0) {
+    packageNames.add(originalPackageName);
```

Comment:
Package can be overridden in the version meta which we don't parse at the moment (version 5.x has this info hashed) so we add this explicitly if not found otherwise:

The following line says that despite dependency is typescript@4.8.2 we will use 4.8.4 so that would end up being the only installed version:
```
  /@phenomnomnominal/tsquery@4.1.1(typescript@4.8.4):
    resolution: {integrity: sha512-jjMmK1tnZbm1Jq5a7fBliM4gQwjxMU7TFoRNwIyzwlO+eHPRCFv/Nv+H/Gi1jc3WR7QURG8D5d0Tn12YGrUqBQ==}
    peerDependencies:
      typescript: ^3 || ^4
    dependencies:
      esquery: 1.5.0
      typescript: 4.8.2
```